bin/tx-prover/src/proxy/mod.rs

```diff
@@ -154,77 +142,73 @@ impl ProxyHttp for LoadBalancer {
     }
 
     /// Returns [HttpPeer] corresponding to the worker that will handle the current request.
     ///
-    /// Here we select the next worker from the pool in a round-robin fashion. We then add the
-    /// request to the worker's queue and wait until it gets to the front of it. Then, we construct
-    /// and return the [HttpPeer]. The peer is configured with timeouts, and HTTP/2.
+    /// Here we enqueue the request and wait for it to be at the front of the queue and a worker
+    /// becomes available. We then set the SNI, timeouts, and enable HTTP/2.
     ///
     /// Note that the request is not removed from the queue here. It will be returned later in
     /// [Self::logging()] once the worker processes the it.
     async fn upstream_peer(
         &self,
-        session: &mut Session,
+        _session: &mut Session,
         ctx: &mut Self::CTX,
     ) -> Result<Box<HttpPeer>> {
-        if ctx.tries > 0 {
-            Self::remove_request_from_queue(Self::get_request_id(session)?).await;
-        }
+        let request_id = ctx.request_id;
 
-        // Select the worker in a round-robin fashion
-        let worker = self.lb.select(b"", 256).ok_or(Error::new_str("Worker not found"))?;
-
-        // Read request ID from headers
-        let request_id = {
-            let id = Self::get_request_id(session)?;
-            id.to_string()
-        };
-
-        // Enqueue the request ID in the worker queue
-        // We use a new scope to release the lock after the operation
+        // Add the request to the queue.
         {
-            let mut ctx_guard = QUEUES.write().await;
-            let worker_queue = ctx_guard.entry(worker.clone()).or_insert_with(Vec::new);
-
-            // Limit queue length
-            if worker_queue.len() >= self.max_queue_items {
-                Self::create_queue_full_response(session).await?;
-            }
-
-            worker_queue.push(request_id.to_string());
+            let mut queue = QUEUE.write().await;
+            queue.push(request_id);
         }
 
         // Wait for the request to be at the front of the queue
         loop {
-            // We use a new scope for each iteration to release the lock
+            // We use a new scope for each iteration to release the lock before sleeping
             {
-                let ctx_guard = QUEUES.read().await;
-                if let Some(worker_queue) = ctx_guard.get(&worker) {
-                    if worker_queue[0] == request_id {
-                        break;
-                    }
-                } else {
-                    return Err(Error::new_str("Worker not found"));
+                let queue = QUEUE.read().await;
+                // The request is at the front of the queue.
+                if queue[0] != request_id {
+                    continue;
+                }
+
+                // Check if there is an available worker
+                if let Some(worker) = self.get_available_worker().await {
```

Comment:
Maybe it could be done separately, but it might be good to add some `info!()` logs to output information about the workers here (whether all workers are busy, if an available worker was found, etc.). Additionally, maybe when a request comes in we can log the state of the queue, etc.